### PR TITLE
Sscs 11431 Tribs

### PR DESCRIPTION
**Pull request by nickhill111**
https://tools.hmcts.net/jira/browse/SSCS-11431

This ticket also takes code from [SSCS-11667](https://tools.hmcts.net/jira/browse/SSCS-11667)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/resources/config/application.properties [hunk 1]
```diff
@@ -130,7 +130,7 @@ feature.gaps-switchover.enabled: ${GAPS_SWITCHOVER_FEATURE:false}
 feature.snl.enabled: ${SCHEDULE_LISTING_FEATURE:false}
 feature.snl.adjournment.enabled: ${ADJOURNMENT_FEATURE:false}
 feature.case-access-management.enabled: ${CASE_ACCESS_MANAGEMENT_FEATURE:false}
-feature.postHearings.enabled: ${POST_HEARINGS_FEATURE:false}
+feature.postHearings.enabled: ${POST_HEARINGS_FEATURE:true}
```
Comment (amoslv1) [inline src/main/resources/config/application.properties:133]: revert

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceMidEventHandlerTest.java [hunk 2]
```diff
@@ -45,7 +45,7 @@ public class ActionFurtherEvidenceMidEventHandlerTest {
 
     @Before
     public void setUp() {
-        handler = new ActionFurtherEvidenceMidEventHandler(footerService);
+        handler = new ActionFurtherEvidenceMidEventHandler(footerService, true);
```
Comment (amoslv1) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceMidEventHandlerTest.java:48]: Shouldn't this be false for the majority of the tests? Then reassign handler with true on the post hearing test(s)

**Review by amoslv1 — CHANGES_REQUESTED** [review]
Some questions

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceMidEventHandlerTest.java (hunk 2 above)
Comment (nickhill111) [reply src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceMidEventHandlerTest.java]: changed to true

File: src/main/resources/config/application.properties (hunk 1 above)
Comment (nickhill111) [reply src/main/resources/config/application.properties]: done

**Review by amoslv1 — APPROVED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/sscs/functional/handlers/actionfurtherevidence/ActionFurtherEvidenceAboutToStartHandlerTest.java [hunk 3]
```diff
@@ -53,6 +53,7 @@ public void givenAboutToStartCallback_shouldSetItemsInFurtherActionDropdownMenu(
             .assertThat().body("data.furtherEvidenceAction.list_items", hasItem(hasEntry("code", "informationReceivedForInterlocTcw")))
             .assertThat().body("data.furtherEvidenceAction.list_items", hasItem(hasEntry("code", "sendToInterlocReviewByJudge")))
             .assertThat().body("data.furtherEvidenceAction.list_items", hasItem(hasEntry("code", "sendToInterlocReviewByTcw")))
+            .assertThat().body("data.furtherEvidenceAction.list_items", hasItem(hasEntry("code", "sendToInterlocReviewByTcw")))
```
Comment (amoslv1) [inline src/functionalTest/java/uk/gov/hmcts/reform/sscs/functional/handlers/actionfurtherevidence/ActionFurtherEvidenceAboutToStartHandlerTest.java:56]: this is a duplicate of the line above

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceAboutToStartHandler.java [hunk 4]
```diff
@@ -47,14 +49,14 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
...
         sscsCaseData.setFurtherEvidenceAction(new DynamicList(listOptions.get(0), listOptions));
     }
 
-    private List<DynamicListItem> getFurtherActionEvidenceItems(List<FurtherEvidenceActionDynamicListItems> items) {
-        return items.stream()
-            .filter(item -> !item.equals(ADMIN_ACTION_CORRECTION) || isPostHearingsEnabled)
+    private List<DynamicListItem> getFurtherActionEvidenceItems(SscsCaseData sscsCaseData) {
+        return Stream.of(values())
+            .filter(item -> !item.equals(ADMIN_ACTION_CORRECTION)
+                || (!GAPS.equals(sscsCaseData.getSchedulingAndListingFields().getHearingRoute()) && isPostHearingsEnabled))
```
Comment (amoslv1) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceAboutToStartHandler.java:59]: This is missing test coverage
image

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceAboutToSubmitHandler.java [hunk 5]
```diff
@@ -174,10 +176,13 @@ public static void checkWarningsAndErrors(SscsCaseData sscsCaseData, ScannedDocu
             && scannedDocument != null && scannedDocument.getValue() != null
             && scannedDocument.getValue().getOriginalSenderOtherPartyId() != null) {
             if (!sscsCaseData.getOriginalSender().getValue().getCode().equalsIgnoreCase(OTHER_PARTY.getCode() + scannedDocument.getValue().getOriginalSenderOtherPartyId())) {
-                preSubmitCallbackResponse
-                    .addError("The PDF evidence does not match the Original Sender selected");
+                preSubmitCallbackResponse.addError("The PDF evidence does not match the Original Sender selected");
             }
         }
+
+        if (isPostHearingsEnabled && isGapsCase(sscsCaseData) && isPostHearingRequest(sscsCaseData)) {
```
Comment (amoslv1) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceAboutToSubmitHandler.java:183]: Could do with more comprehensive test coverage
image

**Review by amoslv1 — CHANGES_REQUESTED** [review]
Could use some extra code coverage as those test files are already hovering around 90% (bit low)

**Review by jarekPierchala — APPROVED** [review]

**Review by amoslv1 — APPROVED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/sscs/functional/handlers/actionfurtherevidence/ActionFurtherEvidenceAboutToStartHandlerTest.java (hunk 3 above)
Comment (nickhill111) [reply src/functionalTest/java/uk/gov/hmcts/reform/sscs/functional/handlers/actionfurtherevidence/ActionFurtherEvidenceAboutToStartHandlerTest.java]: removed

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceAboutToStartHandler.java (hunk 4 above)
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceAboutToStartHandler.java]: added unit tests for CC

**Review by amoslv1 — APPROVED** [review]

**Review by amoslv1 — DISMISSED** [review]

**Review by DGordonV1 — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceAboutToSubmitHandler.java [hunk 6]
```diff
@@ -292,30 +298,42 @@ private static boolean isOriginalSenderDwp(SscsCaseData sscsCaseData) {
...
             .anyMatch(doc -> doc.getValue() != null && isNotBlank(doc.getValue().getType())
                 && documentType.getValue().equals(doc.getValue().getType()));
     }
 
+    private static boolean isPostHearingRequest(SscsCaseData sscsCaseData) {
+        return isSetAsideApplication(sscsCaseData)
+            || isCorrectionApplication(sscsCaseData)
+            || isStatementOfReasonsApplication(sscsCaseData)
+            || isLibertyToApplyApplication(sscsCaseData)
+            || isPermissionToApplyApplication(sscsCaseData);
```
Comment (amoslv1) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceAboutToSubmitHandler.java:312]: I think that these bottom two need to be enclosed in posthearingBEnabled flag

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceMidEventHandler.java [hunk 7]
```diff
@@ -39,11 +40,8 @@ public class ActionFurtherEvidenceMidEventHandler implements PreSubmitCallbackHa
     public static final String ONLY_ONE_POSTPONEMENT_AT_A_TIME = "Only one request for postponement can be submitted at a time";
     public static final String OTHER_PARTY_ORIGINAL_PARTY_ERROR = "You cannot select 'Other party hearing preferences' as a Document Type as an Other party not selected from Original Sender list";
     private final FooterService footerService;
-
-    @Autowired
-    public ActionFurtherEvidenceMidEventHandler(FooterService footerService) {
-        this.footerService = footerService;
-    }
+    @Value("${feature.postHearings.enabled}")
+    private final boolean isPostHearingsEnabled;
```
Comment (amoslv1) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceMidEventHandler.java:44]: Is posthearings B flag also required here?

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceMidEventHandlerTest.java [hunk 8]
```diff
@@ -485,6 +486,62 @@ public void givenARequestWithOtherPartyNotSelectedAsOriginalSenderAndOtherPartyH
...
+        ScannedDocument scannedDocument = ScannedDocument.builder()
+            .value(scannedDocDetails)
+            .build();
+
+        sscsCaseData.setScannedDocuments(Collections.singletonList(scannedDocument));
+
+        PreSubmitCallbackResponse<SscsCaseData> response = handler.handle(MID_EVENT, callback, USER_AUTHORISATION);
+
+        assertEquals(0, response.getErrors().size());
+    }
```
Comment (amoslv1) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/furtherevidence/actionfurtherevidence/ActionFurtherEvidenceMidEventHandlerTest.java:542]: One more test for post hearing flag not enabled

**Review by amoslv1 — CHANGES_REQUESTED** [review]
Potentially need post hearing b flag in a few places, just worth double checking

**Review by amoslv1 — APPROVED** [review]

**Review by jarekPierchala — DISMISSED** [review]

**Review by jarekPierchala — APPROVED** [review]

**Review by hajam09 — APPROVED** [review]

**Pull request merged**